Add prefix, suffix, and comments to regexp assemble data files

theseion opened the pull request:
This PR improves regexp-assemble.pl by introducing a bit more flexibility:
- regexp data files can now contain comments (`#`)
- regexp data files can now contain flag comments (`#!`)
- regexp data files can now contain prefix comments (`#^`)
- regexp data files can now contain suffix comments (`#$`)
- empty lines in regexp data files will now be ignored

These changes make it possible to comment individual regular expressions, explaining their purpose and origin. It also allows us to remove all of the comments in the rules files about manual modification of the assembled regular expression (included in the PR).

theseion commented:
@dune73 @lifeforms would you mind taking a look? The approach is a bit opinionated of course, but I think the added flexibility should help a lot.

dune73 commented:
Ah, sucks. I overlooked this PR for the meeting last night. Very sorry.

dune73 commented:
This looks neat. I was expecting comment line options for the assembly call, but this is probably even better.

Am I getting this right: When you say `#` is comment and `#!` is the flag comment, then the flag comment in the file is meant to me `##!`, thus a combination. You may have to to make this more explicit in the README and maybe also add a generic comment to the data files since it is not obvious what this really means when a newbie looks at the data file without knowing the context.

theseion commented:
I updated the format this morning. A comment is now anything prefixed with `##!` and the different types of special comments have a suffix character (more consistent and less likely to clash with regular expressions). Do you still think it needs to be better explained in the readme?

dune73 commented:
Yes, I do.

Take this for example:

---

A line starting with `##!+` can be used to pass global flags to the script. The last found flag comment line overwrites
all previous flag comment lines. The resulting expression will be prefixed with the flags. For example, the two lines
```
#! i
a+b|c
```
---

The explanation and the example are not congruent for me. When you only look at the example / data file, this looks very dangerous. When you look at the README, it's very complicated.

If we could make this intuitive, it would be much easier.

Ideally, there is something like this in every data file:

```
# This is the data file used to generate a regular expression for a CRS rule. 
# The generation of the regular expression happens with the help of 
# util/regexp-assemble/regexp-assemble.pl.
# The ID of the rule in question is part of the file name of this data file. 
# Read more about the format of this data file and the use of the assembly 
# script in util/regexp-assemble/README.md.
#
# The regular expression generated is enriched with a global flag to ... FIXME
# This is defined on the following line (see README for more infos):
##!+ i
```

This is awfully redundant. But it also takes a casual user by the hand when he / she reads a data file for the first time.
The first block of the example is generic, so it can be copied into any data file.

theseion commented:
Yeah, I missed those examples. Thanks.

- I've added an explanation comment to all data files,
- added an example to the readme,
- fixed the existing examples.

Do you think it's better now?

dune73 commented:
Yes, I do. Much clearer now.

May I suggest you add an empty line (or an empty comment line) after the intro of the data files. I just think it looks better and is a bit better to read than with the long comment line and then a full start with a complex regex. No functional difference though.

theseion commented:
This should do it :)

dune73 commented:
It does, it does. Thank you very much for taking the time to polish this. I have not execute the new script or played around with it, but as far as documentation is concerned I think this is a big step forward and I'm entirely satisfied.

theseion commented:
Sweet. Thanks for checking!

franbuehler commented:
[Meeting decision August 2021](https://github.com/coreruleset/coreruleset/issues/2169#issuecomment-891255600): @franbuehler  check this, regenerate and compare the regexes

dune73 commented:
Hey @franbuehler, did you have time to check this and regenerate the regexes for comparison?

franbuehler commented on util/regexp-assemble/README.md:
````diff
@@ -0,0 +1,64 @@
+^#!/bin/bash
+```
+
+This data file would produce the following assebled expression: `(?i)\b(?:^#!\/bin\/bash|--a--|__b__)\W*(`
````
assebled -> assembled

franbuehler commented on util/regexp-assemble/regexp-932210.data:
```diff
@@ -1,3 +1,19 @@
+##! - `##!^`: the prefix comment
+##! - `##!$`: the suffix comment
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
```
Maybe the prefix comment could be added instead of adding it after assembling the rule?
`;\s*\.\s*`

theseion replied:
Missed that one. Fixed.

franbuehler commented on util/regexp-assemble/regexp-933131.data:
```diff
@@ -1,3 +1,19 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
Looks good!

franbuehler commented on util/regexp-assemble/regexp-933160.data:
```diff
@@ -1,3 +1,19 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
Looks good. But maybe a prefix and suffix comment could be added instead of adding them after assembling the rule.

theseion replied:
Missed that one, thanks. Fixed.

franbuehler commented on util/regexp-assemble/regexp-933161.data:
```diff
@@ -1,3 +1,19 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
Looks good! Maybe prefix and suffix comment could be added.

theseion replied:
Missed that one, thanks. Fixed.

franbuehler commented on util/regexp-assemble/regexp-941130.data:
```diff
@@ -1,3 +1,19 @@
+##! - `##!^`: the prefix comment
+##! - `##!$`: the suffix comment
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
```
The generated regex and the old rule differ. I think this is not a problem. It's the same, but we should replace the generated regex in the rule file.

```
$ diff 941130.old 941130.new
1,2c1
< (?i)[\s\S](?:!ENTITY\s+(?:\S+|%\s+\S+)\s+(?:PUBLIC|SYSTEM)|x(?:link:href|html|mlns)|data:text/html|pattern\b.*?=|formaction|\@import|;base64)\b
< 
---
> (?i)[\s\S](?:!ENTITY\s+(?:\S+|%\s+\S+)\s+(?:PUBLIC|SYSTEM)|x(?:link:href|html|mlns)|data:text\/html|pattern\b.*?=|formaction|;base64|@import)\b
```

theseion replied:
Hm. More missing escapes....

theseion replied:
After updating the script this regex is now identical except for an additional escape. I think that's ok.

theseion replied:
I checked the output of the original script again and the additional backslash before `/html` is present there too. So the regex in the rule must have been modified by hand. Both scripts yield the same expression.

franbuehler commented on util/regexp-assemble/regexp-942120.data:
```diff
@@ -1,3 +1,21 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
Looks good! The ignore case flag is written in a new way. This should be replaced in the rule file.
`(?i)` instead of `(?i:...)`

theseion replied:
Done.

franbuehler commented on util/regexp-assemble/regexp-942130.data:
```diff
@@ -1,3 +1,21 @@
+##! - `##!$`: the suffix comment
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
+##!+ i
```
Looks good! The ignore case flag is written in a new way. This should be replaced in the rule file.
`(?i)` instead of `(?i:...)`

theseion replied:
Done.

franbuehler commented on util/regexp-assemble/regexp-942140.data:
```diff
@@ -1,3 +1,21 @@
+##! - `##!$`: the suffix comment
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
+##!+ i
```
The `\b` is missing and the ignore case flag is written in a new way:

```
$ diff 942140.old 942140.new
1c1,2
< (?i:\b(?:(?:m(?:s(?:ys(?:ac(?:cess(?:objects|storage|xml)|es)|(?:relationship|object|querie)s|modules2?)|db)|aster\.\.sysdatabases|ysql\.db)|pg_(?:catalog|toast)|information_schema|northwind|tempdb)\b|s(?:(?:ys(?:\.database_name|aux)|qlite(?:_temp)?_master)\b|chema(?:_name\b|\W*\())|d(?:atabas|b_nam)e\W*\())
---
> (?i)(?:(?:m(?:s(?:ys(?:ac(?:cess(?:objects|storage|xml)|es)|(?:relationship|object|querie)s|modules2?)|db)|aster\.\.sysdatabases|ysql\.db)|pg_(?:catalog|toast)|information_schema|northwind|tempdb)\b|s(?:(?:ys(?:\.database_name|aux)|qlite(?:_temp)?_master)\b|chema(?:_name\b|\W*\())|d(?:atabas|b_nam)e\W*\()
> 

```

theseion replied:
Fixed.

franbuehler commented on util/regexp-assemble/regexp-942150.data:
```diff
@@ -1,3 +1,23 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
Looks good!

franbuehler commented on util/regexp-assemble/regexp-942170.data:
```diff
@@ -1,3 +1,21 @@
+##! - `##!$`: the suffix comment
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
+##!+ i
```
Ignore case flag should be changed in the rule file so that it reflects the result of the assembling script:

```
$ diff 942170.old 942170.new
1c1
< (?i:(?:select|;)\s+(?:benchmark|sleep|if)\s*?\(\s*?\(?\s*?\w+)
---
> (?i)(?:select|;)\s+(?:benchmark|sleep|if)\s*?\(\s*?[(]?\s*?\w+
```

theseion replied:
Done. I also updated the data file as the parenthesis `(` wasn't being escaped but presented as a character class (which doesn't make sense).

franbuehler commented on util/regexp-assemble/regexp-942180.data:
```diff
@@ -1,3 +1,21 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
The new result and the old rule differ. The changes should be reviewed and changed in the rule file if ok.

theseion replied:
Looks like more issues with escapes.

theseion replied:
After updating the script this regex is now identical except for an additional escape. I think that's ok.

theseion replied:
Same as above: the original script includes an additional backslash that is not present in the expression in the rule. Both script versions produce the same expression.

franbuehler reviewed (changes requested):
Hi @theseion 
I started to review this PR and I found some differences between the newly generated regexes and the existing rules.
Could you please have a look at my comments and let me know if I should continue my review or if you want to check the results of the rest of the regexes and compare them to the rule files.
Some differences are ok, like those with the ignore case flag. But these should be reflected in the resulting rule files.

theseion commented:
Wow, that's great, thanks. I'll take a look and let you know later today.

theseion commented on util/regexp-assemble/regexp-941160.data:
```diff
@@ -1,3 +1,21 @@
+##! This is a data file used to generate a regular expression for a CRS rule. 
```
I'd usually argue for a separate PR. However, after making the change to the data file the new script produces the exact regex that is currently in the rule, so I think it's better to roll it into this PR.

theseion commented on util/regexp-assemble/regexp-932130.data:
```diff
@@ -1,3 +1,19 @@
+##! Please refer to util/regexp-assemble/README.md for a full explanation.
+
 \$\(.*\)
 \$\{.*\}
```
Indeed. I will have to look into this.

theseion replied:
I've adjusted the lexing logic in the script to match that of the old version. The produced regular expressions will have more unnecessary escape sequences but should now match the output of the old version.

theseion commented:
@franbuehler You can continue now.

franbuehler commented on rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf:
```diff
@@ -181,8 +181,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_H
-# Note that after assemble an ignore case flag (i) is added to the to the Regexp::Assemble output:
-#   Add ignore case flag between '?' and ':': "(?i:...)"
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:<\w[\s\S]*[\s/]|['\"](?:[\s\S]*[\s/])?)(?:on(?:d(?:e(?:vice(?:(?:orienta|mo)tion|proximity|found|light)|livery(?:success|error)|activate)|r(?:ag(?:e(?:n(?:ter|d)|xit)|(?:gestur|leav)e|start|drop|over)|op)|i(?:s(?:c(?:hargingtimechange|onnect( ...
```
Thank you for fixing the data file (source of this regex) so that it reflects the manually made changes in an old commit. Just the ignore case flag has to be changed. The regex will be the same then, yes.

theseion replied:
Done.

franbuehler commented on util/regexp-assemble/regexp-942380.data:
```diff
@@ -1,3 +1,19 @@
+##! empty lines will be ignored completely.
+##! 
+##! Three special comments are at your disposal to influence the assembled expression:
+##! - `##!+`: the flag comment
```
The ignore case flag is missing in this source file and also in the resulting generated regex.
And please replace the ignore case flag in the resulting file. Also see my comment in the source (data) file.

theseion replied:
Done

franbuehler reviewed (changes requested):
Hi @theseion 

Thanks a lot for your changes, comments and your work!
I reviewed the remaining rules and wrote some comments. Mostly on changing the ignore case flag in the rule files.
But it looks good to me and I think we can soon merge this PR.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -73,9 +73,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_H
-# to the Regexp::Assemble output:
-#   Add ignore case flag and word boundary: "(?i:\bASSEMBLE_OUTPUT)"
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:\b(?:(?:m(?:s(?:ys(?:ac(?:cess(?:objects|storage|xml)|es)|(?:relationship|object|querie)s|modules2?)|db)|aster\.\.sysdatabases|ysql\.db)|pg_(?:catalog|toast)|information_schema|northwind|tempdb)\b|s(?:(?:ys(?:\.database_name|aux)|qlite(?:_temp)?_master)\b|chema(?:_name\b|\W*\())|d(?:atabas|b_nam ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -175,9 +169,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:\b(?:u(?:nion(?:[\w(\s]*?select| select \@)|ser\s*?\([^\)]*?)|(?:c(?:onnection_id|urrent_user)|database)\s*?\([^\)]*?|s(?:chema\s*?\([^\)]*?|elect.*?\w?user\()|into[\s+]+(?:dump|out)file\s*?[\"'`]|from\W+information_schema\W|exec(?:ute)?\s+master\.)|[\"'`](?:;?\s*?(?:union\b\s*?(?:(?:distin|s ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -249,9 +240,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:[\"'`](?:;*?\s*?waitfor\s+(?:delay|time)\s+[\"'`]|;.*?:\s*?goto)|alter\s*?\w+.*?cha(?:racte)?r\s+set\s+\w+))" \
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -320,9 +308,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:;\s*?shutdown\s*?(?:[#;]|/\*|--|\{)|waitfor\s*?delay\s?[\"'`]+\s?\d|select\s*?pg_sleep))" \
```
Escape issue.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -370,9 +355,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:create\s+(?:procedure|function)\s*?\w+\s*?\(\s*?\)\s*?-|;\s*?(?:declare|open)\s+[\w-]+|procedure\s+analyse\s*?\(|declare[^\w]+[@#]\s*?\w+|exec\s*?\(\s*?\@))" \
```
Done.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -399,9 +381,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:;\s*?(?:(?:(?:trunc|cre|upd)at|renam)e|(?:inser|selec)t|de(?:lete|sc)|alter|load)\b\s*?[\[(]?\w{2,}|create\s+function\s+.+\s+returns))" \
```
Done.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -441,9 +420,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:^[\W\d]+\s*?(?:alter\s*(?:a(?:(?:pplication\s*rol|ggregat)e|s(?:ymmetric\s*ke|sembl)y|u(?:thorization|dit)|vailability\s*group)|c(?:r(?:yptographic\s*provider|edential)|o(?:l(?:latio|um)|nversio)n|ertificate|luster)|s(?:e(?:rv(?:ice|er)|curity|quence|ssion|arch)|y(?:mmetric\s*key|nonym)|togro ...
```
Done.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -584,9 +557,6 @@ SecRule ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:not\s+between\s+(?:(?:(?:'[^']*')|(?:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 # Not supported by re2 (backreferences, lookaheads).
 #
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -654,9 +620,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# first non-capturing group from the Regexp::Assemble output:
-#    ASSEMBLE_OUTPUT | s/^(?:/(?i:/
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:[\"'`](?:\s*?(?:(?:between|x?or|and|div)[\w\s-]+\s*?[+<>=(),-]\s*?[\d\"'`]|like(?:[\w\s-]+\s*?[+<>=(),-]\s*?[\d\"'`]|\W+[\w\"'`(])|[!=|](?:[\d\s!=+-]+.*?[\"'`(].*?|[\d\s!=]+.*?\d+)$|[^\w\s]?=\s*?[\"'`])|(?:\W*?[+=]+\W*?|[<>~]+)[\"'`])|(?:/\*)+[\"'`]+\s?(?:/\*|--|\{|#)?|\d[\"'`]\s+[\"'`]\s+\d|whe ...
```
Escaping issue

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -686,9 +649,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:(?:(?:(?:trunc|cre|upd)at|renam)e|(?:inser|selec)t|de(?:lete|sc)|alter|load)\s*?\(\s*?space\s*?\(|,.*?[)\da-f\"'`][\"'`](?:[\"'`].*?[\"'`]|(?:\r?\n)?\z|[^\"'`]+)|\Wselect.+\W*?from))" \
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -718,9 +678,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:(?:n(?:and|ot)|(?:x?x)?or|between|\|\||like|and|div|&&)[\s(]+\w+[\s)]*?[!=+]+[\s\d]*?[\"'`=()]|\d(?:\s*?(?:between|like|x?or|and|div)\s*?\d+\s*?[\-+]|\s+group\s+by.+\()|/\w+;?\s+(?:between|having|select|like|x?or|and|div)\W|--\s*?(?:(?:insert|update)\s*?\w{2,}|alter|drop)|#\s*?(?:(?:insert|up ...
```
Escaping issue

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -746,10 +703,7 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-#    ASSEMBLE_OUTPUT | s/^(?:/(?i:/
+#   ./regexp-assemble.pl regexp-942260.data
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:[\"'`]\s*?(?:(?:n(?:and|ot)|(?:x?x)?or|between|\|\||and|div|&&)\s+[\s\w]+=\s*?\w+\s*?having\s+|like(?:\s+[\s\w]+=\s*?\w+\s*?having\s+|\W*?[\"'`\d])|[^?\w\s=.,;)(]++\s*?[(@\"'`]*?\s*?\w+\W+\w|\*\s*?\w+\W+[\"'`])|(?:union\s*?(?:distinct|[(!@]*?|all)?\s*?[([]*?\s*?select|select\s+?[\[\]()\s\w\.,\"' ...
```
Possible escaping issue

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -776,9 +730,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:\b(?:(?:r(?:egexp|like)|n(?:and|ot)|(?:x?x)?or|like|and|div)\s+\s*?\w+\(|b(?:etween\s+\s*?\w+\(|inary\s*?\(\s*?\d)|cha?r\s*?\(\s*?\d)|\)\s*?when\s*?\d+\s*?then|(?:\|\||&&)\s+\s*?\w+\(|[\"'`]\s*?(?:--|\{|#)|\/\*!\s?\d+))" \
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -805,9 +756,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:[\"'`](?:;\s*?(?:begin|while|if)|[\s\d]+=\s*?\d|\s+and\s*?=\W)|(?:\(\s*?select\s*?\w+|order\s+by\s+if\w*?|coalesce)\s*?\(|\w[\"'`]\s*?(?:(?:[-+=|@]+\s+?)+|[-+=|@]+)[\d(]|[\s(]+case\d*?\W.+[tw]hen[\s(]|\+\s*?\d+\s*?\+\s*?\@|\@\@\w+\s*?[^\w\s]|\W!+[\"'`]\w|\*/from))" \
```
Escaping issue

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -842,9 +790,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:^(?:[\"'`\\\\]*?(?:[^\"'`]+[\"'`]|[\d\"'`]+)\s*?(?:n(?:and|ot)|(?:x?x)?or|between|\|\||like|and|div|&&)\s*?[\w\"'`][+&!@(),.-]|.?[\"'`]$)|\@(?:[\w-]+\s(?:between|like|x?or|and|div)\s*?[^\w\s]|\w+\s+(?:between|like|x?or|and|div)\s*?[\"'`\d]+)|[\"'`]\s*?(?:between|like|x?or|and|div)\s*?[\"'`]?\ ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -873,9 +818,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:[\"'`](?:\s*?(?:is\s*?(?:[\d.]+\s*?\W.*?[\"'`]|\d.+[\"'`]?\w)|\d\s*?(?:--|#))|(?:\W+[\w+-]+\s*?=\s*?\d\W+|\|?[\w-]{3,}[^\w\s.,]+)[\"'`]|[\%&<>^=]+\d\s*?(?:between|like|x?or|and|div|=))|(?i:n?and|x?x?or|div|like|between|not|\|\||\&\&)\s+[\s\w+]+(?:sounds\s+like\s*?[\"'`]|regexp\s*?\(|[=\d]+x)| ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -931,9 +873,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:^[\W\d]+\s*?(?:(?:alter\s*(?:a(?:(?:pplication\s*rol|ggregat)e|s(?:ymmetric\s*ke|sembl)y|u(?:thorization|dit)|vailability\s*group)|c(?:r(?:yptographic\s*provider|edential)|o(?:l(?:latio|um)|nversio)n|ertificate|luster)|s(?:e(?:rv(?:ice|er)|curity|quence|ssion|arch)|y(?:mmetric\s*key|nonym)|to ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -1052,9 +988,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQU
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:\band\b(?:\s+(?:'[^=]{1,10}'(?:\s*?[=<>])?|\d{1,10}(?:\s*?[=<>])?)| ?(?:[\'\"][^=]{1,10}[\'\"]|\d{1,10}) ?[=<>]+))" \
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -966,9 +905,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:[\"'`](?:\s*?(?:(?:\*.+(?:(?:an|i)d|between|like|x?or|div)\W*?[\"'`]|(?:between|like|x?or|and|div)\s[^\d]+[\w-]+.*?)\d|[^\w\s?]+\s*?[^\w\s]+\s*?[\"'`]|[^\w\s]+\s*?[\W\d].*?(?:--|#))|.*?\*\s*?\d)|[()\*<>%+-][\w-]+[^\w\s]+[\"'`][^,]|\^[\"'`])" \
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -1122,9 +1050,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQU
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:xp_(?:reg(?:re(?:movemultistring|ad)|delete(?:value|key)|enum(?:value|key)s|addmultistring|write)|(?:servicecontro|cmdshel)l|e(?:xecresultset|numdsn)|ntsec(?:_enumdomains)?|terminate(?:_process)?|availablemedia|loginconfig|filelist|dirtree|makecab)|s(?:p_(?:(?:addex ...
```
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -1155,9 +1080,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQU
-# to the Regexp::Assemble output:
-#   (?i:ASSEMBLE_OUTPUT)
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:\b(?:(?:s(?:elect\b.{1,100}?\b(?:(?:(?:length|count)\b.{1,100}?|.*?\bdump\b.*)\bfrom|to(?:p\b.{1,100}?\bfrom|_(?:numbe|cha)r)|(?:from\b.{1,100}?\bwher|data_typ)e|instr)|ys_context)|in(?:to\b\W*?\b(?:dump|out)file|sert\b\W*?\binto|ner\b\W*?\bjoin)|u(?:nion\b.{1,100}? ...
```
Rule 942410

theseion replied:
Done

theseion replied:
Done

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -1087,11 +1020,6 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQU
-# is added after the Regexp::Assemble output:
-#   (?i:\bASSEMBLE_OUTPUT\W*?\()
 #
 SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:\b(?:c(?:o(?:n(?:v(?:ert(?:_tz)?)?|cat(?:_ws)?|nection_id)|(?:mpres)?s|ercibility|(?:un)?t|alesce)|ur(?:rent_(?:time(?:stamp)?|date|user)|(?:dat|tim)e)|h(?:ar(?:(?:acter)?_length|set)?|r)|iel(?:ing)?|ast|r32)|s(?:t(?:d(?:dev(?:_(?:sam|po)p)?)?|r(?:_to_date|cmp))|u(?:b( ...
```
Rule 942410.
Done

theseion commented:
Thanks for the amazing work!

There still appear to be some escape issues which I will have to investigate.

theseion commented:
I've looked into the issue with the bad escape sequences. The cause of something like `\(?` turning into `\(\?` is the first fix I made due to removed escapes. The problem appears when using the faster `_fastlex` subroutine, using `_lex` instead solves the issue.

I will revert back to using `_lex` only as it produces more reliable results. The one side-effect this has is that some characters may be escaped even though they don't need to be. I am willing to pay that price.

@franbuehler do you have a script on hand to compare the output produced by the scripts? If not I will probably write one. That could then even be used to test that generated expressions aren't modified in the rule directly and we can put that into the verification pipeline.

dune73 commented:
Trying to catch up here. The unnecessary escapes: Is that a new problem or was it present with the old assembly script already?

theseion commented:
> Trying to catch up here. The unnecessary escapes: Is that a new problem or was it present with the old assembly script already?

Correct. The original script produced expressions that could contain unnecessary backslashes. I had made a change to my implementation to produce expressions that would align more closely with the original expressions to simplify the transition. I had not realized at the time that the old script could also produce faulty expressions (which my "fix" reintroduced).

dune73 commented:
I see. Good thinking.

How viable do you think it is to actually get this done?

Also: It's probably nice2have, but also something that can be postponed via a new issue for future fixing.

theseion commented:
> I see. Good thinking.
> 
> How viable do you think it is to actually get this done?

Removing the additional escapes? Not easily. It requires a fix to Regexp::Assemble.

> 
> Also: It's probably nice2have, but also something that can be postponed via a new issue for future fixing.

theseion commented:
Oh, you obviously meant the script :) That will definitely take some work. I'll open an issue once I have a simple version for verification, then we can do the rest there.

dune73 commented:
Sounds like a plan. And I generally like our trajectory of cleaning up the ruleset, one step at a time. We've come a long way since CRS 2.2.x.

theseion commented on rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:
```diff
@@ -146,11 +143,8 @@ SecRule REQUEST_BASENAME|REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIE
-#   (?i:ASSEMBLE_OUTPUT)
 #
-SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i:(?:select|;)\s+(?:benchmark|sleep|if)\s*?\(\s*?\(?\s*?\w+)" \
+SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i)(?:select|;)\s+(?:benchmark|sleep|if)\s*?\(\s*?\(?\s*?\w+" \
```
Fixed with update to regexp-assemble.pl.

theseion commented:
@franbuehler I'm done.

There's a new script called `regexp-compare.py`. Simply run it to go over all of the regex files and compare the output to the current rule.
If you want to compare the output of the script to a previous version of the rules (e.g. before my changes), simply check out the rules of that version before running the script, e.g.
```bash
git checkout HEAD~10 rules/\*.conf
```

Important: I generated the new expressions with the fix to `regexp-assemble.pl` that I made in the other PR. So before running `regexp-compare.py`, make sure that you have that version checked out:
```bash
git remote add theseion git@github.com:theseion/coreruleset.git
git fetch theseion
git checkout 2210-dont-cook-hex-escapes
git submodule update --init
git checkout <whatever branch you want to work on>
git checkout 2210-dont-cook-hex-escapes util/regexp-assemble/regexp-assemble.pl
```

Most of the changes you'll see are either missing flags or removed / added escapes. Here's a list of some things to look out for:
- `\/` is equivalent to `/`
- `\@` is equivalent to `@`
- `x|y` is equivalent to `[xy]`, which is an optimization that `_lex` does
- rule 942330 looks very different, until you realize that all that happened is this (simplyfied): `[a]*(?:[b]|[c])` -> `[a]*[b]|[a]*[c]`. Those two expressions are semantically equivalent
- I've updated some data files and replaced single character character classes, mostly `[(]` to `\(`. This is what caused the escaping issue that you had noticed previously. You must have known of this problem when you deconstructed the regular expressions, right? Otherwise, I don't see why one would use a single character character class... :)

Let me know if you have any questions.

franbuehler commented:
I had a look at this PR again. There are some differences between the newly generated regexes and the current ones (current v3.4/dev branch). But most of the new rules just have a new order of the alternatives for example rule 942190 or the special characters are written and organised in a new way for example in rule 942300. Or xy|xz becomes to x(y|z) or vice versa like in 942330. 942380 is a bit difficult but also there it should mean the same. 
All these changes are due to changes in the underlying data files.

I think enough work and review was put into this PR and it seems ready to be merged.

Thank you @theseion for your work and your endurance!!

franbuehler commented:
[Meeting decision Oct 4](https://github.com/coreruleset/coreruleset/issues/2197#issuecomment-933754113): ready to be merged after the rebase that should fix the conflicts.

theseion commented:
@fzipi I've rebase on top of v3.4/dev. Now the tests fail because no web server is running...?

theseion commented:
@franbuehler There's still an open change request from you. Can I perform the merge?

franbuehler commented:
I'll check the change request.
